java-squid/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -172,13 +178,116 @@ ProgramState put(Symbol symbol, SymbolicValue value) {
+      }
+      newReferences = increaseReference(newReferences, value);
+      PMap<Symbol, SymbolicValue> newValues = values.put(symbol, value);
+      return new ProgramState(newValues, newReferences, constraints, visitedPoints, stack);
     }
     return this;
   }
 
+  private static PMap<SymbolicValue, Integer> decreaseReference(PMap<SymbolicValue, Integer> givenReferences, SymbolicValue sv) {
+    Integer value = givenReferences.get(sv);
+    Preconditions.checkNotNull(value);
+    return givenReferences.put(sv, value - 1);
+  }
+
+  private static PMap<SymbolicValue, Integer> increaseReference(PMap<SymbolicValue, Integer> givenReferences, SymbolicValue sv) {
+    Integer value = givenReferences.get(sv);
+    if (value == null) {
+      return givenReferences.put(sv, 1);
+    } else {
+      return givenReferences.put(sv, value + 1);
+    }
+  }
+
+  ProgramState remove(Symbol symbol) {
+    if (symbol.isUnknown()) {
+      return this;
+    }
+    SymbolicValue symbolicValue = values.get(symbol);
+    if (symbolicValue != null) {
+      PMap<Symbol, SymbolicValue> newValues = values.remove(symbol);
+      PMap<SymbolicValue, Integer> newReferences = decreaseReference(references, symbolicValue);
+      PMap<SymbolicValue, Object> newConstraints = constraints;
+      if (newReferences.get(symbolicValue) == 0 && isDisposable(symbolicValue, newConstraints.get(symbolicValue)) && !inStack(symbolicValue)) {
+        newConstraints = newConstraints.remove(symbolicValue);
+        newReferences = newReferences.remove(symbolicValue);
+      }
+      return new ProgramState(newValues, newReferences, newConstraints, visitedPoints, stack);
+    }
+    return this;
+  }
+
+  private static boolean isDisposable(SymbolicValue symbolicValue, @Nullable Object constraint) {
+    return SymbolicValue.isDisposable(symbolicValue) && (constraint == null || !(constraint instanceof ObjectConstraint) || ((ObjectConstraint) constraint).isDisposable());
+  }
+
+  private boolean inStack(SymbolicValue symbolicValue) {
+    for (SymbolicValue value : stack) {
+      if (symbolicValue == value) {
+        return true;
+      } else if (value instanceof SymbolicValue.BinarySymbolicValue) {
+        return inBinarySymbolicValue(symbolicValue, (SymbolicValue.BinarySymbolicValue) value);
+      }
+    }
+    return false;
+  }
+
+  private static boolean inBinarySymbolicValue(SymbolicValue symbolicValue, SymbolicValue.BinarySymbolicValue value) {
+    Deque<SymbolicValue.BinarySymbolicValue> queue = new ArrayDeque<>();
+    queue.addLast(value);
+    while (!queue.isEmpty()) {
+      SymbolicValue.BinarySymbolicValue last = queue.removeLast();
+      for (SymbolicValue current : ImmutableList.of(last.leftOp, last.rightOp)) {
+        if (current instanceof SymbolicValue.BinarySymbolicValue) {
+          queue.addLast((SymbolicValue.BinarySymbolicValue) current);
+        } else if (current == symbolicValue) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)